File: python/234_palindromeLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} head
    # @return {boolean}
    def isPalindrome(self, head):
        if head is None or head.next is None:
            return True
        
        fast = head
        slow = head
        pre = None
        
        while fast and fast.next:
            pre = slow
            slow = slow.next
            fast = fast.next.next
        
        l1 = head
        l2 = pre.next
        pre.next = None
        l2 = self.reverseList(l2)
        
        while l1 and l2:
            if l1.val != l2.val:
                return False
            l1 = l1.next
            l2 = l2.next
        return True
    def reverseList(self, head):
        cur_next = None
        prev = None
        while head:
            cur_next = head.next
            head.next = prev
            prev = head
            head = cur_next
        return prev

File: leetcode/cpp_solutions/234_palindromeLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (head == NULL || head->next == NULL){
            return true;
        }

        ListNode* fast = head, *slow = head, *pre = NULL;
        while(fast && fast->next){
            pre = slow;
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* l1 = head;
        ListNode* l2 = pre->next;
        pre->next = NULL;
        l2 = reverseList(l2);
        while(l2 && l1){
            if (l2->val != l1->val){
                return false;
            }
            l1 = l1->next;
            l2 = l2->next;
        }
        return true;
    }

    ListNode* reverseList(ListNode* head) {
        ListNode* cur_next = NULL;
        ListNode* prev = NULL;
        while(head != NULL){
            cur_next = head->next;
            head->next = prev;
            prev = head;
            head = cur_next;
        }
        return prev;
    }
};
